proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

```diff
@@ -0,0 +1,177 @@
+}
+func CFunctionReturningAForwardDeclaredInterface1() -> ForwardDeclaredInterface!
+func CFunctionTakingAForwardDeclaredInterface1(_ param: ForwardDeclaredInterface!)
+func CFunctionReturningAForwardDeclaredProtocol1() -> ForwardDeclaredProtocol!
+func CFunctionTakingAForwardDeclaredProtocol1(_ param: ForwardDeclaredProtocol!)
+```
+
+More usage examples can be found in these tests introduced here: https://github.com/apple/swift/pull/61606
+
+## Detailed design
+
+Modifications lie almost exclusively in the ClangImporter -- specifically in the SwiftDeclConverter. If asked to convert a
+`clang::ObjCInterfaceDecl` or `clang::ObjCProtocolDecl` with no definition the SwiftDeclConverter will now return a placeholder type instead of
+bailing. These placeholder types are as described:
+
+```swift
+// @class Foo turns into
+@available(*, unavailable, message: “This Objective-C class has only been forward declared; import its owning module to use it”)
+class Foo : NSObject {}
+
+// @protocol Bar turns into
+@available(*, unavailable, message: “This Objective-C protocol has only been forward declared; import its owning module to use it”)
+protocol Bar : NSObjectProtocol {}
+```
+
+Permitted usages of these types are intentionally limited. You will be able to use Objective-C and C declarations that refer to these types without issue.
+You will be able to pass around instances of these incomplete types from Swift to Objective-C and vice versa.
+You’ll also be able to call the set of methods provided by NSObject or NSObjectProtocol on instances.
+
+Using the type itself directly in Swift is forbidden by the attached unavailable attribute. This is to keep the impact of the change small and to prevent unsound declarations,
+such as declaring a new Swift class that inherits from or conforms to such a type. You will also not be able to create new instances of these types in Swift.
+
+To make these limitations more intuitive to users, a new diagnostic has been added. This diagnostic helps guide users when they attempt to access a member on the incomplete synthesized
+type, expecting the complete type. The new diagnostics are shown in the last 4 lines of this example:
+
+```swift
+swift-client-fwd-declared.swift:6:7: error: value of type 'Foo' has no member 'sayHello'
+myFoo.sayHello()
+~~~~~ ^~~~~~~~
+<unknown>:0: note: class 'Foo' is a placeholder for a forward declared Objective-C interface and may be missing members; import the definition to access the complete interface
+foo-bar-consumer.h:3:1: note: interface 'Foo' forward declared here
+@class Foo;
+^
+```
+
+The feature is gated behind a new frontend flag `-enable-import-objc-forward-declarations`. This flag is on by default for Swift version 6 and onwards. The flag is always disabled in the REPL, as
+allowing it leads to confusing behavior. The [existing ClangImporter diagnostics](https://forums.swift.org/t/pitch-improved-clangimporter-diagnostics/52687) will still
```

Comment:
Can you elaborate a bit on what the "confusing behavior" is, so that readers have a better understanding of why it's not supported there?